tempodb/encoding/iterator_multiblock.go

```diff
@@ -108,32 +128,70 @@ func (i *multiblockIterator) iterate(ctx context.Context) {
+			// is found, simply combine it with that check.
+			if currentMerger == nil || foundLowerID {
+				// currentMerger is uninitialized, or a new lowerID was found in the iterators
+				// so reset the currentMerger to use the new object. The second case shouldn't
+				// happen after merging for another ID has happened, so track it just in case.
+				metricMultiBlockIteratorWorkDropped.WithLabelValues(i.level).Add(float64(workCount))
+				workCount = 0
+				currentMerger = i.mergerFactory(currentID, currentObject)
+
+				// Track the ID for comparisons in further iterations of this loop.
+				lowestID = currentID
 
-			if comparison == 0 {
-				lowestObject, _ = i.combiner.Combine(i.dataEncoding, currentObject, lowestObject)
+				// The current bookmark isn't cleared since there is potential to find a
+				// lower ID down the list. If that were to happen after clearing this bookmark,
+				// there would be data loss.
+				bookmarkToClear = b
+			} else if shouldMerge {
+				// currentID matches the lowestID, so merge in the new object
+				_, err := currentMerger.Merge(i.dataEncoding, currentObject)
+				if err != nil {
+					// TODO: log? the existing code assumes it'll work
+				}
+				workCount++
 				b.clear()
-			} else if len(lowestID) == 0 || comparison == -1 {
-				lowestID = currentID
-				lowestObject = currentObject
-				lowestBookmark = b
 			}
 		}
 
-		if len(lowestID) == 0 || len(lowestObject) == 0 || lowestBookmark == nil {
-			i.err.Store(errors.New("failed to find a lowest object in compaction"))
+		// This means that all blocks being iterated were at EOF, otherwise the return
+		// call when checking the iterators for errors would have already dropped out.
+		if currentMerger == nil {
+			return
+		}
+
+		// Get the final merged result.
+		mergedObject, err := currentMerger.Result(i.dataEncoding)
+		if err != nil {
+			i.err.Store(fmt.Errorf("failed to serialize merged result object: %w", err))
 			return
 		}
 
-		// Copy slices allows data to escape the iterators
+		// The array backing the currentID slice is still owned by the underlying iterator
+		// buffer, so it must be copied out, otherwise when the result is read, the buffer
+		// will most likely be filled with new data. The mergedObject however is already escaped
+		// by the result call, and by virtue of being deserialized and then serialized back to
+		// bytes, so it can be used directly.
 		res := iteratorResult{
-			id:     append([]byte(nil), lowestID...),
-			object: append([]byte(nil), lowestObject...),
+			id:     append([]byte(nil), currentMerger.ID()...),
+			object: mergedObject,
 		}
 
-		lowestBookmark.clear()
+		// Now that this was definitely the object being worked on, the first bookmark for the
+		// current object can be cleared. The nil check is not technically required, since
+		// if it is nil, currentMerger must also be nil and this code path will not be hit,
+		// but do so anyways for sanity and safety.
+		if bookmarkToClear != nil {
```

Comment:
This assumes that we didn't see a lower ID twice. If we did, we'd fail to clear the first one and we'd try to merge it again. I guess, strictly speaking, this is OK - we wouldn't drop data, we'd just end up with duplicate data, which the querier can already handle. It still feels like we're jumping through hoops to end up in a somewhat dubious state.